### PR TITLE
#82 create record for config

### PR DESCRIPTION
**Pull request by TimoKramer**
hi there,
maybe you can check my solution on the config issue? It is now thought that one has to load the config-namespace to create a config. I did that to reuse the connector-code and not break the old ways to handle configuration.
I also updated the config-docs...
best regards, Timo

P.S: I would squash first before you merge so please tell me if you are happy with this solution or what would be better.

### CODE REVIEW

**Comment by TimoKramer** [comment]
I improved the solution so that you don't have to load config-namespace and run load-config to use create/connect/delete/exist?-functions. Now the config var is queryable and config is loaded on start. with a reload-config function you can now update the config with args otherwise it is using the environ/systemprops.

File: project.clj [hunk 1]
```diff
@@ -8,11 +8,13 @@
...
                  [persistent-sorted-set     "0.1.2"]
                  [org.clojure/tools.reader "1.3.2"]
+                 [environ "1.1.0"]
                  [io.replikativ/hitchhiker-tree "0.1.5"]
                  [io.replikativ/superv.async "0.2.9"]
                  [io.lambdaforge/datalog-parser "0.1.1"]]
 
-  :plugins [[lein-cljsbuild "1.1.7"]]
+  :plugins [[lein-cljsbuild "1.1.7"]
+            [lein-environ "1.1.0"]]
```
Comment (TimoKramer) [inline project.clj]: das mache ich wieder raus

File: src/datahike/config.cljc [hunk 2]
```diff
@@ -1,26 +1,89 @@
...
-(s/def ::host string?)
-(s/def ::port int?)
-(s/def ::uri string?)
+(s/def ::backend #{:mem :file :level :pg})
+(s/def ::username (s/nilable string?))
+(s/def ::password (s/nilable string?))
+(s/def ::path (s/nilable string?))
+(s/def ::host (s/nilable string?))
+(s/def ::port (s/nilable int?))
+(s/def ::uri (s/nilable string?))
```
Comment (TimoKramer) [inline src/datahike/config.cljc]: that's to be removed

File: src/datahike/config.cljc [hunk 3]
```diff
@@ -1,26 +1,89 @@
...
 
 (s/def ::schema-on-read boolean?)
 (s/def ::temporal-index boolean?)
 
-(s/def :datahike/config (s/keys :req-un [::backend]
-                                :opt-un [::host ::port ::path ::username ::password]))
+(s/def :datahike/store (s/keys :req-un [::backend]
+                               :opt-un [::username ::password ::path ::host ::port]))
+(s/def :datahike/schema (s/keys :opt-un [::schema-on-read ::temporal-index]))
```
Comment (kordano) [inline src/datahike/config.cljc:20]: namespaces will be changed in the future. So we would have something like `:index/schema-flexibility` with better naming.

Comment (TimoKramer) [reply src/datahike/config.cljc:20]: should I make schema-on-read and temporal-index first level options? Means taking schema-record out.

**Comment by TimoKramer** [comment]
FMPOV this PR is good to go.

**Comment by whilo** [comment]
This looks in general good to me, but I am not an expert in configuration management. The main thing that I see missing (besides the missing SSL encryption of issue #133) are tests that check that when the configuration is done wrong a proper error is raised.

**Comment by TimoKramer** [comment]
You're right @whilo. I will insert logging-statements commented out and implement logging as next issue. I will try to add postgresql-ssl-support as well but as a separate issue.

**Comment by TimoKramer** [comment]
squashed and ready to merge

**Comment by TimoKramer** [comment]
one more thing I was thinking about now regarding the postgresql-config:
Wouldn't it be better to have `:path` for `:file` and `:level` and a `:database`-key for postgresql? I mean having path for a postgresql-database is kind of hackish.

**Comment by kordano** [comment]
You are right about these specific backend keys. I would solve that with a separate PR.

**Pull request merged**